string.c

```diff
@@ -9159,6 +9159,84 @@ rb_str_end_with(int argc, VALUE *argv, VALUE str)
     return Qfalse;
 }
 
+/*
+ *  call-seq:
+ *     str.delete_prefix!(substr)   -> self or nil
+ *
+ *  Deletes leading <code>substr</code> from <i>str</i>, returning
+ *  <code>nil</code> if no change was made.
+ *
+ *     "hello".delete_prefix!("hel")  #=> lo
+ *     "hello".delete_prefix!("llo")  #=> nil
+ */
+
+static VALUE
+rb_str_delete_prefix_bang(VALUE str, VALUE substr)
+{
+    rb_encoding *enc;
+    char *strptr, *substrptr, *s;
+    long olen, substrlen, len;
+
+    str_modify_keep_cr(str);
+    StringValue(substr);
+    if (is_broken_string(substr)) return Qnil;
+    rb_enc_check(str, substr);
+
+    // return if not start_with with substr
+    substrlen = RSTRING_LEN(substr);
+    if (substrlen <= 0) return Qnil;
+    olen = RSTRING_LEN(str);
+    if (olen < substrlen) return Qnil;
+    strptr = RSTRING_PTR(str);
+    substrptr = RSTRING_PTR(substr);
+    if (memcmp(strptr, substrptr, substrlen) != 0) return Qnil;
```

Comment:
Ah, now I use rb_str_drop_bytes, so VALUE prefix was not required, and long prefixlen was enough as you say. I fixed.